src/v/pandaproxy/rest/proxy.cc

```diff
@@ -9,52 +9,125 @@
 
 #include "pandaproxy/rest/proxy.h"
 
+#include "config/property.h"
+#include "config/rest_authn_endpoint.h"
+#include "net/unresolved_address.h"
 #include "pandaproxy/api/api-doc/rest.json.h"
 #include "pandaproxy/logger.h"
+#include "pandaproxy/parsing/exceptions.h"
+#include "pandaproxy/parsing/from_chars.h"
 #include "pandaproxy/rest/configuration.h"
 #include "pandaproxy/rest/handlers.h"
+#include "pandaproxy/types.h"
+#include "redpanda/request_auth.h"
+#include "utils/gate_guard.h"
 
 #include <seastar/core/future-util.hh>
 #include <seastar/core/memory.hh>
 #include <seastar/http/api_docs.hh>
 
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/constants.hpp>
+#include <boost/algorithm/string/split.hpp>
+
+#include <algorithm>
+#include <string>
+
 namespace pandaproxy::rest {
 
+namespace {
+net::unresolved_address address_from_url(std::string_view url) {
+    std::vector<std::string> parts;
+    boost::split(parts, url, boost::is_any_of("/:"), boost::token_compress_on);
+
+    // After split, there should be 4 strings in vector
+    // [http, domain, port, trailing route]. Anything else
+    // is an unacceptable url.
+    if (parts.size() != 4) {
+        throw parse::error(parse::error_code::not_acceptable, "Invalid URL");
+    }
+
+    auto res = parse::from_chars<uint16_t>{}(parts[2]);
+    if (res.has_error()) {
+        throw parse::error(
+          parse::error_code::not_acceptable, "INT conversion fail");
+    }
+
+    return net::unresolved_address{parts[1], res.value()};
+}
+} // namespace
+
 using server = ctx_server<proxy>;
 
-server::routes_t get_proxy_routes() {
+template<typename Handler>
+auto wrap(configuration& cfg, Handler h) {
+    return [&cfg, _h{std::move(h)}](
+             server::request_t rq,
+             server::reply_t rp) -> ss::future<server::reply_t> {
+        auto h{_h};
```

Comment:
We are now able to use the technique described here: https://github.com/scylladb/seastar/blob/master/doc/lambda-coroutine-fiasco.md

Reply:
@BenPope thanks for the reference. i was staring at this for a while, until i realized that wrap here is returning a lambda coroutine. but the the technique in the document you referenced, IIUC, is a technique applied at a callsite (like sending a wrapper into a then(lambda) continuation).

so i think it allows `h` to be removed here from the stack, but all the call sites would need to be updated, right?

perhaps we can tackle a bunch of these conversions across the tree separately so that this PR doesn't expand too much in scope?

Reply:
Good point, I don't think the technique is required here for a couple of reasons, but primarily there's no co_await before use of h.

Reply:
Instead of using a wrapper on the handlers, let's put authentiction within the handler_adaptor in server.cc